src/gui/qgscheckablecombobox.h

```diff
@@ -0,0 +1,238 @@
+#include <QStyledItemDelegate>
+
+#include "qgis_gui.h"
+
+class QEvent;
+
+/** \class QgsCheckableItemModel
+ * \ingroup gui
+ * QStandardItemModel subclass which makes all items checkable
+ * by default.
+ * \since QGIS 3.0
+ * \note not available in Python bindings
+ **/
+#ifndef SIP_RUN
+class QgsCheckableItemModel : public QStandardItemModel
+{
+    Q_OBJECT
+
+  public:
+
+    /** Constructor for QgsCheckableItemModel.
+     * \param parent parent object
+     */
+    QgsCheckableItemModel( QObject *parent = nullptr );
+
+    /** Returns a combination of the item flags: items are enabled
+     * (ItemIsEnabled), selectable (ItemIsSelectable) and checkable
+     * (ItemIsUserCheckable).
+     * \param index item index
+     */
+    virtual Qt::ItemFlags flags( const QModelIndex &index ) const;
+
+    /** Returns the data stored under the given role for the item
+     * referred to by the index.
+     * \param index item index
+     * \param role data role
+     */
+    virtual QVariant data( const QModelIndex &index, int role = Qt::DisplayRole ) const;
+
+    /** Sets the role data for the item at index to value.
+     * \param index item index
+     * \param value data value
+     * \param role data role
+     * \returns true on success, false otherwise
+     */
+    virtual bool setData( const QModelIndex &index, const QVariant &value, int role = Qt::EditRole );
+
+  signals:
+
+    /** This signal is emitted whenever the items checkstate has changed.
+     */
+    void itemCheckStateChanged();
+};
+
+
+/** \class QgsCheckBoxDelegate
+ * \ingroup gui
+ * QStyledItemDelegate subclass for QgsCheckableComboBox. Needed for
+ * correct drawing of the checkable items on Mac and GTK.
+ * \since QGIS 3.0
+ * \note not available in Python bindings
+ **/
+
+class QgsCheckBoxDelegate : public QStyledItemDelegate
+{
+    Q_OBJECT
+
+  public:
+
+    /** Constructor for QgsCheckBoxDelegate.
+     * \param parent parent object
+     */
+    QgsCheckBoxDelegate( QObject *parent = nullptr );
+
+    /** Renders the delegate using the given painter and style option
+     * for the item specified by index.
+     * \param painter painter to use
+     * \param option style option
+     * \param index item index
+     */
+    virtual void paint( QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index ) const;
+};
+#endif
+
+/** \class QgsCheckableComboBox
+ * \ingroup gui
+ * QComboBox subclass which allows selecting multiple items.
+ * \since QGIS 3.0
+ **/
+
+class GUI_EXPORT QgsCheckableComboBox : public QComboBox
+{
+    Q_OBJECT
+
+    Q_PROPERTY( QString separator READ separator WRITE setSeparator )
+    Q_PROPERTY( QString defaultText READ defaultText WRITE setDefaultText )
+    Q_PROPERTY( QStringList checkedItems READ checkedItems WRITE setCheckedItems )
+
+  public:
+
+    /** Constructor for QgsCheckableComboBox.
+     * \param parent parent object
+     */
+    QgsCheckableComboBox( QWidget *parent = nullptr );
```

Comment:
Technically speaking this should be 

     QgsCheckableComboBox( QWidget *parent SIP_TRANSFERTHIS = nullptr );

But I don't think it makes any real world difference